### ISSUE TITLE
[Feat]: SSR support

### ISSUE THREAD

**Issue by bartlomiejzuber** [davidjerleke/embla-carousel#202]
> [!NOTE]  
> **Edit:** This feature is under development:
>
> - #1095
>
> _Edited by @davidjerleke 2024-12-26._

---

### Bug is related to

- [x] The Embla Carousel core (all versions)

### Embla Carousel version

- v4.5.3

### Describe the bug

- When carousel is being rendered via SSR there's noticeable flickering effect on first slide on each page refresh (regardless of cache).

### CodeSandbox

- https://codesandbox.io/s/suspicious-burnell-mfs1q

### Steps to reproduce

1. Go to https://mfs1q.sse.codesandbox.io/
2. Refresh the page and look at carousel.

### Expected behavior

- It shouldn't flicker when SSR is used.

### Additional context

![flicker](https://user-images.githubusercontent.com/9931428/122062513-c58a8700-cdef-11eb-995c-6e3d4de82726.gif)

**Comment by davidcetinkaya** [comment]
Hi @bartlomiejzuber,

Thank you for your bug report. This might come as a surprise but this is expected. Embla Carousel is a client side script and won't do anything on the server. This is because Embla needs to read the container and slide dimensions in order to align the carousel correctly, and place some slides in loop positions. This has to be done on the client because element dimensions doesn't exist on the server. So this is what's happening:

- You serve your HTML and CSS from the server.
- The client script mounts and Embla picks up the container and slide dimensions (slide sizes, spacing between slides etc).
- Embla aligns the slides and positions slides needed to create the loop effect.
- Hence, you'll see the flicker when this happens.

One way to solve this would be to hide the carousel with `opacity: 0;` until Embla has initialized. Alternatively, you can add `opacity: 0;` to all the slides if you still want to display the carousel arrows and dots until Embla has initialized. There are probably other ways to do it but I think you get the idea.

I hope this helps.

Best,
David

**Comment by manojVivek** [comment]
@davidcetinkaya Do you think the approach suggested in this issue, helps fix this problem? https://github.com/FormidableLabs/nuka-carousel/issues/786#issue-928155497 ?

**Comment by davidcetinkaya** [comment]
Hi @manojVivek,

Currently, the HTML gets server side rendered and you don't loose any important SEO. I think the flicker thing is solvable, but if this is going to be implemented we need to ask ourselves at what cost? How much will it affect bundle size and what would be a good solution for this?

**Comment by mariaaraujo** [comment]
@davidjerleke Thanks for mentioning this issue! But could you help me with something... How could I control if the embla has initialized or not?

**Comment by davidjerleke** [comment]
@mariaaraujo which Embla package are you using?

**Comment by MartinCura** [comment]
Hi all, in case of any help to anyone, i have a page with a big carousel (product images) of which the currently selected is stored in a query param of the URL. To achieve a server-rendered HTML that is already pointing to the correct image i added a translation to the `<div>` container of the items and which is child of the element that holds the reference.

Reduced example (from my components based on shadcn's):

```tsx
  <div ref={carouselRef} className="overflow-hidden">
    <div
      ref={ref}
      className="flex"
      style={{ transform: `translateX(-${initialIdx * 100}%)` }}
    />
  </div>
```

This translation is overridden by Embla with a `transform3d` ([here](https://github.com/davidjerleke/embla-carousel/blob/e32119ab938111e5594a5c71ef14ab9bea09359a/packages/embla-carousel/src/components/Translate.ts#L20) if i'm not mistaken) as soon as it loads/hydrates client-side (and becomes interactive), but there's no flicker because it matches up. If you have any margin between the items in the carousel, you may have to adjust for that.

Hope it helps anyone!

**Comment by davidjerleke** [comment]
This feature has been added to the roadmap:

- #841

**Comment by davidjerleke** [comment]
@sarussss, thank you for your question. The SSR feature will be opt-in. By default, it won’t be active. So if that’s your concern, you will be able to use Embla just as you do today without any problems.

But the bundle size will grow a little bit because SSR is the kind of feature that has to be built into the core. However, the bundle size won’t grow significantly. 

Did that answer your question?

**Comment by davidjerleke** [comment]
Hey @sarussss, 

Sorry if I wasn’t clear 🙂. The SSR feature needs to be integrated into the core because it impacts several parts of the core code, so it can't be a plugin.

What I meant was that I might consider removing a different feature and turning that into a plugin instead, in order to reduce the core library's bundle size.

**Comment by davidjerleke** [comment]
@sarussss no, **SSR can’t be a plugin**. It has to be added to the core library, meaning the `embla-carousel` package because of technical reasons.

**Comment by MartinCura** [comment]
SSR support incoming? 💥 Please link the docs when it's ready to try it, thank you so much.

**Comment by davidjerleke** [comment]
@MartinCura thank you for your interest in the SSR feature. This feature will be part of version 9, the  next major release. At this time, I don’t have an ETA for the first release candidate, as there is still significant work to be done. I appreciate your patience and enthusiasm as we continue to make progress.

**Comment by intelligence** [comment]
Any progress on this? :)

**Comment by davidjerleke** [comment]
@intelligence it’s done, will be released in v9:

- #841

**Comment by intelligence** [comment]
That's awesome! Any rough ETA on when it might be released?

**Comment by davidjerleke** [comment]
> That's awesome! Any rough ETA on when it might be released?

@intelligence unfortunately, no. Since this is OSS without decent funding, I’m unable to commit to ETAs.

**Comment by intelligence** [comment]
> [@intelligence](https://github.com/intelligence) unfortunately, no. Since this is OSS without decent funding, I’m unable to commit to ETAs.

I totally understand that! Looking forward to testing out the new release when it's out. Thanks for all your hard work!

**Comment by davidjerleke** [comment]
@intelligence side note — love your profile pic!

**Comment by parsagholipour** [comment]
@davidjerleke Can you give us access to it before the release of v9, in npm@next tag maybe

**Comment by davidjerleke** [comment]
@parsagholipour I understand the request, but context switching is extremely ineffective (and costly) in my situation. Since I only have ~2h/week available for v9 development, adding extra steps like publishing interim builds to npm@next would take away from making progress toward the actual release. For now, my focus is on getting v9 finished and shipped.

**Comment by nickdash777** [comment]
@davidjerleke hey, any update on v9 release? Thanks!

**Comment by davidjerleke** [comment]
@nickdash777 @parsagholipour @intelligence @MartinCura, 

`v9.0.0-rc01` has now been released with SSR support:

- [Link to release](https://github.com/davidjerleke/embla-carousel/releases/tag/v9.0.0-rc01)
- [SSR docs](https://www.embla-carousel.com/guides/server-side-rendering/)

**Comment by MartinCura** [comment]
> @MartinCura, 
> 
> `v9.0.0-rc01` has now been released with SSR support:

Awesome! Just today i had seen the rc.0. Excited about trying it soon.

Are any upgrade steps necessary?

**Comment by davidjerleke** [comment]
@MartinCura thanks. Please follow the link to the release in my [previous comment](https://github.com/davidjerleke/embla-carousel/issues/202#issuecomment-3786444164) and you’ll see the release notes and migration guide.
